chasm/lib/callback/executor.go

```diff
@@ -0,0 +1,219 @@
+package callback
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"go.temporal.io/server/chasm"
+	callbackspb "go.temporal.io/server/chasm/lib/callback/gen/callbackpb/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	commonnexus "go.temporal.io/server/common/nexus"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/service/history/queues"
+	"go.uber.org/fx"
+)
+
+// HTTPCaller is a method that can be used to invoke HTTP requests.
+type HTTPCaller func(*http.Request) (*http.Response, error)
+type HTTPCallerProvider func(queues.NamespaceIDAndDestination) HTTPCaller
+
+// InvocationTaskExecutor is responsible for the invocation of a callback.
+// For Nexus callbacks this will be an HTTP call, for other callbacks this
+// could be any CHASM invocation.
+// This is the CHASM port of the HSM taskExecutor.executeInvocationTask functionality
+// from components/callbacks/executors.go.
+type InvocationTaskExecutor struct {
+	InvocationTaskExecutorOptions
+}
+
+type InvocationTaskExecutorOptions struct {
+	fx.In
+
+	Config             *Config
+	MetricsHandler     metrics.Handler
+	Logger             log.Logger
+	NamespaceRegistry  namespace.Registry
+	HTTPCallerProvider HTTPCallerProvider
+	HTTPTraceProvider  commonnexus.HTTPClientTraceProvider
+	HistoryClient      resource.HistoryClient
+}
+
+func NewInvocationTaskExecutor(opts InvocationTaskExecutorOptions) *InvocationTaskExecutor {
+	return &InvocationTaskExecutor{
+		InvocationTaskExecutorOptions: opts,
+	}
+}
+
+func (e *InvocationTaskExecutor) Execute(
+	ctx context.Context,
+	invokerRef chasm.ComponentRef,
+	taskAttributes chasm.TaskAttributes,
+	task *callbackspb.InvocationTask,
+) error {
+	var ns *namespace.Namespace
+	// var invoker *Invoker
+	var callback *Callback
+
+	// Read the invoker component and load invocation arguments.
+	_, err := chasm.ReadComponent(
+		ctx,
+		invokerRef,
+		func(c *Callback, ctx chasm.Context, _ any) (struct{}, error) {
```

Comment:
Just return what you need here instead of cloning the entire proto IMHO. I typically would rather not cloning components for use outside of the chasm contexts.